Automattic/woocommerce-payments
# Fixes Pay for Order page functionality in split-UPE

Fixes #5808 

#### Changes proposed in this Pull Request


# Pay for Order Page: Preserved at Last

Previously when the split-UPE was enabled, the Pay for Order page had very limited functionality. The card payment method could fail, if the customer elected to change their currency, and the other additional payment methods would fail in all cases. I have identified three issues that are causing these frustrating failures.

1. If order currency and customer selected currency do not match, if the customer selected currency is not supported by available payment method, payment intent will produce an error and payment elements will fail to mount.
2. Pay for order form incorrectly detects when WCPay payment method is selected at checkout.
3. Payment intent confirms successfully, but `process_redirect_payment` function exits early, order is not updated, and instead marked as a failure.

I hurdle the first obstacle by ensuring that when the Pay for Order page is open, multi-currency will disregard the currency selected by the customer and enforce that the correct currency on the order is used instead. This ensures that we only present the correct relevant and valid payment methods to the customer on this page and ensures that the customer cannot produce an invalid combination of selected currencies and payment methods. 

I vault the second hurdle by removing the `isWCPayChosen()` function from the Pay for Order client script, instead relying on the `getSelectedUPEGatewayPaymentMethod()` function to test when a WCPay payment method is selected. `isWCPayChosen()` only checks for the presence of the `woocommerce_payments` gateway, ignoring the additional split-UPE gateways, and causing the Pay for Order scripts to consequently be skipped when an alternate WCPay gateway is selected.

The third barrier is the one described in #5854 and should be fixed by #6170.







#### Testing instructions


1. Create an order from WC Admin (WooCommerce > Orders > Add new).
2. Open the newly created order, add an item, customer, billing, and shipping details. Update the order
3. Select the "Customer payment page" to be sent to the Pay for Order page.
4. Fill in payment information and complete checkout for order. Use all payment methods (CC + additional UPE LPMs). 
5. Try to change the currency using the currency switcher widget. Currency displayed on Pay for Order page should remain the same. Payment elements should all load when appropriate without any errors.
6. In all above cases, customer should be able to complete checkout without receiving errors. Order status should be successfully updated to "Processing". Order details on payment intent should all be correct and valid.



-------------------


**Post merge**
